File: components/mutations/EditTweet.ts
import { commitMutation, Environment, graphql } from "react-relay"

import { EditTweetMutationResponse } from "@generated/EditTweetMutation.graphql"

const mutation = graphql`
  mutation EditTweetMutation($input: EditTweetInput!) {
    editTweet(input: $input) {
      tweetGroup {
        id
        tweets {
          body
          mediaURLs
        }
      }
    }
  }
`

interface EditTweetInput {
  id: string
  tweets: {
    body: string
    mediaURLs: readonly string[]
  }[]
}

function getOptimisticResponse(
  input: EditTweetInput
): EditTweetMutationResponse {
  return {
    editTweet: {
      tweetGroup: {
        ...input,
      },
    },
  }
}

export function editTweet(
  environment: Environment,
  input: EditTweetInput
): void {
  const variables = { input }

  commitMutation(environment, {
    mutation,
    variables,
    optimisticResponse: getOptimisticResponse(input),
  })
}

File: ios/Courier/CourierNext/Mutations/EditTweet.swift
import Relay
import RelaySwiftUI

private let mutation = graphql("""
mutation EditTweetMutation($input: EditTweetInput!) {
  editTweet(input: $input) {
    tweetGroup {
      id
      tweets {
        body
        mediaURLs
      }
    }
  }
}
""")

extension Mutation.Mutator where Operation == EditTweetMutation {
    func commit(_ input: EditTweetInput) {
        commit(
            variables: .init(input: input),
            optimisticResponse: [
                "editTweet": [
                    "tweetGroup": [
                        "id": input.id,
                        "tweets": input.tweets.map { tweet in
                            [
                                "body": tweet.body,
                                "mediaURLs": tweet.mediaURLs ?? [],
                            ]
                        }
                    ]
                ]
            ]
        )
    }
}
